### ISSUE TITLE
Spawning new `MaterialMeshBundle`s takes ~3-5x longer in Bevy 0.12 with WASM + WebGL2

### ISSUE THREAD

**Issue by dannymcgee** [bevyengine/bevy#10696]
## Bevy version

* Original: 0.11.3
* Current: 0.12.0

## Relevant system information

Chrome 119
Rust 1.73.0 (Stable)
AMD Ryzen 9 5900X
NVIDIA GeForce RTX 4070

## What's performing poorly?

Spawning a new mesh (and/or a new material? Not sure exactly what the trigger is) that hasn't been rendered previously triggers a significant hiccup in WASM builds with the WebGL2 backend. This was also the case previously, but 0.12 has significantly increased the duration of these "hiccups".

### Repro

The problem should be pretty apparent with any sufficiently complex glTF scene, but I'm also seeing the issue with programmatic geometry (e.g. `Mesh::from(shape::*)`) and simple flat-color materials (e.g. `StandardMaterial { base_color: Color::RED, unlit: true, ..default() }`). 

## Before and After Traces

* Before: https://drive.google.com/file/d/1LChOxM5_1ezh56WPtAVTRXy3pZlkJWex/view?usp=sharing
* After: https://drive.google.com/file/d/1OrA03YPNWpJEKiQlbSLvhmaLHA6mVcDE/view?usp=sharing

## Additional information

The traces above were captured in an app that loads a fairly complex glTF scene. After loading, I click a few buttons which trigger some new (programmatic) meshes to spawn. The same steps were followed in the same order to produce each trace, and the only substantive difference between the two builds was upgrading Bevy from 0.11.3 to 0.12.0.

The specific function call that seems to be triggering the slowdown is `WebGl2RenderingContext::get_program_parameter`, which is called by `glow::web::Context::get_program_link_status`, which is called by `wgpu::Device::create_render_pipeline`. I suspect this is because the shaders have gotten much larger with the introduction of the deferred pipeline, but I'm very much _not_ an expert on this type of thing.

This is a private project for my employer, so I can't share a link to the repo or anything, but I could likely get permission to share a link to the staging site in a DM if that would be helpful.

**Comment by JMS55** [comment]
The stuttering is due to the driver compiling shaders for your GPU. You'll have the same problem on native, it's just that newer graphics APIs have much better compilers that aren't as slow (but still slow enough to cause visible stutter). You ideally want to spawn an instance of each of your materials ahead of time in a loading screen in order to get the pipelines (shaders) cached. You can despawn the material afterwards.

As for why it got worse in Bevy 0.12, try adding the component `ShadowFilteringMethod::Hardware2x2` to your camera.

**Comment by dannymcgee** [comment]
That's super helpful, thanks for the info!

> You'll have the same problem on native, it's just that newer graphics APIs have much better compilers that aren't as slow (but still slow enough to cause visible stutter).

I assume that means that switching to WebGPU would also improve things?

> You ideally want to spawn an instance of each of your materials ahead of time in a loading screen in order to get the pipelines (shaders) cached.

Any insight on which of the `StandardMaterial` properties require a unique shader permutation? Also (this one's a long shot), once a shader is compiled, is there any way we could retrieve and store it on the user's machine so it doesn't need to be recompiled on subsequent application launches?

> As for why it got worse in Bevy 0.12, try adding the component `ShadowFilteringMethod::Hardware2x2` to your camera.

That did help quite a bit, thank you!

**Comment by JMS55** [comment]
> Any insight on which of the StandardMaterial properties require a unique shader permutation? 

For StandardMaterial specifically, take a look at `StandardMaterialKey`. Each possible value of the key type will be a different pipeline. However, that's just one layer of the pipeline (specialization) system. There are other keys that influence the final pipeline. The specific mesh you're using as well as view-specific things like SSAO change the pipeline. Look at `MeshPipelineKey`. Various effects or other systems also have multiple shader permutations, such as SSAO or TAA themselves.

> I assume that means that switching to WebGPU would also improve things?

Probably? I don't have any hard data

>  is there any way we could retrieve and store it on the user's machine so it doesn't need to be recompiled on subsequent application launches?

Not at the moment, no.
